HW1/P6/P6.py

```diff
@@ -0,0 +1,78 @@
+
+import pyspark
+from pyspark import SparkContext
+sc = SparkContext()
+from collections import Counter
+import random
+
+
+text = sc.textFile('shakespeare.txt')
+# parse text and filter certain words
+w1 = text.flatMap(lambda l: l.split(" ")).filter(lambda x: x != '')
+w2 = w1.filter(lambda x: x.isdigit() == False)
+w3 = w2.filter(lambda x: x.isupper() == False)
+word = w3.filter(lambda x: x.endswith('.') == False and x[:-1].isupper() == False)
+
+# build an rdd of the form ((word1, word2), word3)
+r1 = word.zipWithIndex()
+r2a = r1.map(lambda x: (x[1], x[0]))
+r2b = r1.map(lambda x: (x[1]+1, x[0]))
+r2c = r1.map(lambda x: (x[1]+2, x[0]))
+j1 = r2a.leftOuterJoin(r2b).map(lambda x:( x[0], (x[1][1], x[1][0])))
+j2 = j1.leftOuterJoin(r2c).map(lambda x:(x[0], (x[1][1], x[1][0])))
+r2 = j2.filter(lambda x: x[0] >= 2 and x[0] <= 754954)
+r3 = r2.map(lambda x: x[1])
+r4 = r3.map(lambda x: ((x[0], x[1][0]), x[1][1])) #((word1,word2), word3)
+r5 = r4.groupByKey().mapValues(list)
+
+# count the number of word3s
+def make_count(x):
+    C = Counter(x)
+    l = [[k,]*v for k,v in C.items()] # [[w1,w1],[w2,w2,w2]]
+    wcount = [(x[0], len(x)) for x in l]
+    return wcount
```

Comment:
You can map each triplet to 1, and then use reduce to count the number of appearances of each triplet.